### DISCUSSION TITLE
Nullable verification on Expressions?

### DISCUSSION THREAD

**Discussion by mariodvm** [dotnet/csharplang#3567] [Language Ideas]
Take a look at the following warning:
```CSharp
var query = jobs.AsQueryable().OrderBy(j => j.Responsible.Name);
```
and `Job` is defined as:
```CSharp
public class Job 
{ 
    public Contact? Responsible { get; set; }
}
```
So the C# compiler warns in the `OrderBy` because `Responsible` could be null. What it is expected to do in order to remove warning? _"An expression tree lambda may not contain a null propagating operator"_
This is just an Expression, this would not generate any `NullReferenceException` after `IQueryProvider` implementation of many ORMs process it by translating to SQL.

**I propose to review the verification of reference-nullability for `Expression<Func<...>>`**

**Comment by svick** [comment]
> What it is expected to do in order to remove warning?

[I believe you should use the null-forgiving operator, that one works in expression trees:](https://sharplab.io/#v2:C4LgbgpgTglgZgTwDQBMQGoA+ABADAAmwEYAWAbgFgAoPQogOgBkYA7AR0quoGIWBXADYCAhgCMBEfBBZiJ1atgBM+AMLUA3tXzbCJfAFkAFMQDMAHgBSAe1EA+fACsbAZwCUWnZqo6f+MMKh8Nj5oBHwAXkcXegBBZwBFEKgEWQhDV3oAeSgUaAAhBEMHCPsHegAlCGcABysWZxhxCABCegA5YQBbCFdOHwBfakGuGhNCZRU64GEAY2AND21sMeICDu78dXwAcwhgMnxnPYP+iPwAInPOYYUV5WtRfA0n7x1l1SnZ4AB+fEqauoNJqbHbHQ5g4b9IA=)

```c#
var query = jobs.AsQueryable().OrderBy(j => j.Responsible!.Name);
```

> I propose to review the verification of reference-nullability for `Expression<Func<...>>`

What do you expect to be the outcome of that review? That nullability is completely ignored in expression trees? Or something else?

**Reply by atari-monk** [reply]
I dont understand this mess with nulls that is there in .net core 6.0
All i know this removes anoying warns in ef orderby, and its the only way i know

**Comment by mariodvm** [comment]
Nullability propagation is in fact an error on expression trees:
_"An expression tree lambda may not contain a null propagating operator"_
To be consistent, I expect that in case of Expressions there are not warnings on possible nulls.
Consider that Expressions are usually used only to define the expression tree with different objectives than executing the expression itself. e.g. translate to SQL. So if the target language (like SQL) does not have a compatible resource to the reference-nullability, maybe it introduce another mismatch between them. In other words: will this null-forgiving operator being expressed in the tree as another Expression Type node for this kind of expressions too, in order to be considered (or not) by the QueryProviders when translating?

**Comment by svick** [comment]
> In other words: will this null-forgiving operator being expressed in the tree as another Expression Type node for this kind of expressions too, in order to be considered (or not) by the QueryProviders when translating?

It won't be represented in the tree in any way. `j.Responsible!.Name` and `j.Responsible.Name` generate the same `Expression`, the only difference is in the warning.

**Comment by mariodvm** [comment]
> > In other words: will this null-forgiving operator being expressed in the tree as another Expression Type node for this kind of expressions too, in order to be considered (or not) by the QueryProviders when translating?
> 
> It won't be represented in the tree in any way. `j.Responsible!.Name` and `j.Responsible.Name` generate the same `Expression`, the only difference is in the warning.

Exactly, the operand is completely unnecessary. It is supposed to be used to say that 'you' are completly sure that it will not be null: [_you use the null-forgiving operator to declare that expression x of a reference type isn't null_](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/operators/null-forgiving), but this is not the case; the runtime values, after translation to SQL (in DB) can be actually null, but ORMs/SQL already deals with that. If you say:
```CSharp
var query = dbContext.Jobs.Where(j => j.Responsible!.Name.StartsWith("R"));
```
You could be confusedly thinking that every Responsible is not null and if not (for consistency) an exception is expected to be thrown in runtime; but this is not what is happening because QueryProviders will generates SQL such that jobs with null Responsibles will not be in the results, but it will not return an error if some of them are null. 
On the other hand, it is in fact the null propagation what is consistent with the actual behavior:
```CSharp
var query = dbContext.Jobs.Where(j => j.Responsible?.Name.StartsWith("R"));
```
But this is what is not allowed in Expressions. 
As I see, null-forgiving in this case only reinforces explicitly that confusing runtime behavior by muting the warning.

**Comment by HaloFour** [comment]
@mariodvm 

> but this is not the case; the runtime values, after translation to SQL (in DB) can be actually null, but ORMs/SQL already deals with that.

C# has no idea that this is the case.  The compiler doesn't know how any specific queryable provider will handle the expression, so it has to err on the side of caution.\*  If the queryable provider compiles the expression tree and executes it then it can definitely result in a `null` and a `NullReferenceException`.

IMO, ideally, C# would add support for null conditional operators to expression trees.  See #2545.

\* Well, I guess it wouldn't _have_ to.  C# could assume that most people using expression trees are using EF or the like.  NRTs already make a lot of concessions in the sake of practicality.

**Comment by jnm2** [comment]
This seems to be an existing longstanding problem. To get that behavior, the fix is to write:

```cs
var query = dbContext.Jobs.Where(j =>
    j.Responsible == null ? null : j.Responsible.Name.StartsWith("R"));
```

Depending on how your query provider translates this expression, this might be effectively the same thing as `!.`

**Comment by CyrusNajmabadi** [comment]
> As I see, null-forgiving in this case only reinforces explicitly that confusing runtime behavior 

Note: that runtime behavior is *outside* of C# entirely.  You happen to be using an ORM that works that way.  However, C# has no idea that that's the case.  It warns as if you intend your code in the expression to work as the rest of C# does.

**Comment by mariodvm** [comment]
Because Expression Trees are designed to work side by side with QueryProviders from the very beginning, I think this is not so outside of C#. Let's see, C# compiler already assumes the conversion from lamda syntax to Expression Trees as part of the compiler grammar:
```CSharp
Expression<Func<...>> exp = x => ... ;
```
So there is no difference between two line of codes using `Where` extensions:
```CSharp
Jobs.Where(j => j.Responsible?.Name == "John");
```
This is a transparent overload resolution: it will use `Where(this IQueryable<...>, Expression<Func<...>>` or `Where(this IEnumerable<...>, Func<...>)` depending on `Jobs` formal type, so it is expected that new features are so aligned. For example, if compiler warns from nulls on expression within queries, it is also expected that nullable references are also aligned with db model. In the following `Job` definition:
```CSharp
public class Job {
    public string? Description { get; set;}
    public Person Responsible { get; set;}
}
```
We agree that "?" on `Description` should be the direct annotation saying this property is nullable and `Responsible` is not; and this is why the compiler is warning about; but this also means that the underlying DB column **must** also be nullable according with "?". Isn't it? If so, it is expected that the nullable indicators has a corresponding Expression Tree node type in order to make use of them on fluent apis; so `IsRequired` extensions should not longer be required in the following model building:
```CSharp
modelBuilder.Entity<Job>()
            .Property(x => x.Responsible);
            // .IsRequired(); NO LONGER NEEDED BECAUSE RESPONSIBLE IS NOT NULLABLE
```
> The expression `x.Responsible` here is a good example of using Expressions for metadata request only. No protection from runtime evaluation is actually needed in these cases.
>
In the other hand, if the property type were a nullable **value** type, the inspection is actually possible on the Expression tree because the type will be `Nullable<...>`, but there is not a corresponding `NotNullable<...>` for not-nullable reference types.
It could be a long discussion and this is why I'm asking for a review, so that C# stays aligned with the original LINQ goals accordingly.

**Comment by svick** [comment]
@mariodvm 

> so `IsRequired` extensions should not longer be required in the following model building

If you're asking for an easy way for a library like EF to differentiate between `string?` and `string`, then the issue about that is https://github.com/dotnet/runtime/issues/29723.

**Comment by mariodvm** [comment]
> @mariodvm
> 
> > so `IsRequired` extensions should not longer be required in the following model building
> 
> If you're asking for an easy way for a library like EF to differentiate between `string?` and `string`, then the issue about that is [dotnet/runtime#29723](https://github.com/dotnet/runtime/issues/29723).

Yes, indeed, but not only that. I think that a complete review is needed in everything related with nullability and linq: from CLR types like `NotNullable<...>`, also metadata reflection info like: `PropertyInfo.IsNullable`, etc, but also compiler results on query inspection for Expressions like it is discussed here. As a result, everything should be consistent and makes cense in the end.

**Comment by HaloFour** [comment]
@mariodvm 

> Yes, indeed, but not only that. I think that a complete review is needed in everything related with nullability and linq

It was asked before, but what exactly are you looking to come of such a review?  The design of NRTs isn't going to change.  Tools like EF understanding the nullability metadata attributes is outside of C#.  The only thing I could see being considered is whether or not the compiler will bother to warn on possible null dereferencing within an expression tree, but I wouldn't expect that to change.

**Comment by mariodvm** [comment]
> @mariodvm
> 
> > Yes, indeed, but not only that. I think that a complete review is needed in everything related with nullability and linq
> 
> It was asked before, but what exactly are you looking to come of such a review? The design of NRTs isn't going to change. Tools like EF understanding the nullability metadata attributes is outside of C#. The only thing I could see being considered is whether or not the compiler will bother to warn on possible null dereferencing within an expression tree, but I wouldn't expect that to change.

I'm not proposing any specific feature, nor changing what it will no change, I only ask for stand and review what we got and what else it should be done to reach general consistency of nullability and linq, if linq is still considered inside of C#, I suppose that so I expect to consider those things discussed here and also those discussed on [dotnet/runtime#29723](https://github.com/dotnet/runtime/issues/29723) and those on [#2545](https://github.com/dotnet/csharplang/issues/2545). I hope this definitely answer that question.
> NOTE: As you can see it is not my single request and it looks like EF are not so outside of C# as you said, because there are so many people using EF by making queries through C#/linq.

**Comment by HaloFour** [comment]
@mariodvm 

> I'm not proposing any specific feature, nor changing what it will no change, I only ask for stand and review what we got and what else it should be done

Without a list of specific grievances I don't see what would come of that.

> if linq is still considered inside of C#

LINQ is a part of C#, including the generation of expression trees, but the part that translates/executes expression trees into SQL or anything else is not.  C# has no idea as to what EF does or that it can handle `null` in this manner.  Until recently EF allowed for client-side evaluation of parts of a query, where `null`s could matter.  With the runtime support for compiling an expression tree into a delegate which can be invoked, `null`s definitely matter.  The compiler can't know.

**Comment by mariodvm** [comment]
> @mariodvm
> 
> > I'm not proposing any specific feature, nor changing what it will no change, I only ask for stand and review what we got and what else it should be done
> 
> Without a list of specific grievances I don't see what would come of that.
> 
> > if linq is still considered inside of C#
> 
> LINQ is a part of C#, including the generation of expression trees, but the part that translates/executes expression trees into SQL or anything else is not. C# has no idea as to what EF does or that it can handle `null` in this manner. Until recently EF allowed for client-side evaluation of parts of a query, where `null`s could matter. With the runtime support for compiling an expression tree into a delegate which can be invoked, `null`s definitely matter. The compiler can't know.

this part of discussion is going nowhere. I hope other readers understand what is the original point. sorry

**Comment by HaloFour** [comment]
@mariodvm 

> this part of discussion is going nowhere. I hope other readers understand what is the original point. sorry

I'm not sure they do, as multiple people have asked what a "complete review" means.  Without concrete grievances, goals or proposals there's nothing to "review" here.  The team isn't going to act on hand-wavey declarations that "something might not be right".

The only "grievance" that I could distill from your post is that the warning on the expression isn't helpful because there's little you can do aside from using the "null forgiveness" operator to silence it.  I can totally understand that concern and that's something concrete that can be discussed.

**Comment by CyrusNajmabadi** [comment]
>  because there's little you can do aside from using the "null forgiveness" operator to silence it. 

You can also do what @jnm2  suggested here: `j.Responsible == null ? null : j.Responsible.Name.StartsWith("R"));`

It's not great, but it works and has been supported since expression-trees were introduced.

**Comment by IS4Code** [comment]
The fact that such an expression could be translated to SQL should not concern the compiler. The intent is clear - you want to access a member of a potentially null object, which is signalized to the compiler by using `Contact? Responsible` in the first place. It does not matter that the expression is not executed; normal lambdas also don't have to be executed, and, conversely, expressions can be compiled and executed.

If you don't want the warning, define `Contact Responsible` if it is not nullable.

**Comment by mariodvm** [comment]
> The fact that such an expression could be translated to SQL should not concern the compiler. The intent is clear - you want to access a member of a potentially null object, which is signalized to the compiler by using `Contact? Responsible` in the first place. It does not matter that the expression is not executed; normal lambdas also don't have to be executed, and, conversely, expressions can be compiled and executed.
> 
> If you don't want the warning, define `Contact Responsible` if it is not nullable.

Nop, if the property is defined as nullable `Contact? Responsible`, it should keep like that, and if it is used on queries like `jobs.Where(j => j.Responsible.Name == "...")` then it is correct to be warned, but then (for consistency) it should not be an error to query with null propagation on lambda expressions to solve the warning as it should be: `jobs.Where(j => j.Responsible?.Name == "...")` 
Null-forgiveness operator is not an aceptable option either, it is a patch. 
I mean: No patch with "!" please, no warning when it cannot be solved please, and of course: no parching by changing the nullability on the design model, please.
My code right now is full of warnings because of this and I cannot accept to change the db model, nor to patch them all queries with "!" just to remove those wrong warnings in Expressions, because they are wrong: those properties can be null anyway, and just because the compiler don't let me use null propagation in that case I should no patch all the code. 
But yes, all of this is happening because I use "?" on the properties of entities, and it is no so wrong to expect that ORMs could handle that on Fluent API automatically, but again to make it possible, the compiler maybe have to generate Expression types according with it and metadata Type and PropertyInfo objects should "know" about the nullability of properties.
What is the solution? As you can see maybe there is no so simple as a single proposal, so I ask for stand and review all features concerning nullability and linq in order to make it consistent.
And again: _the fact that an expression could be translated to SQL should not concern the compiler_ is not 100% correct. It is the foundation of Linq and QueryProviders in first place, and it do matter with compiler on the Expression production in grammar. 
This is not a crussade against nullability. On the contrary: it is a call to try to make it better for the language and also for keep having a  good language tool to deal with Object Relational Mapping that is why Linq was created in first place.

**Comment by mariodvm** [comment]
...but ok, if it is needed to propose something, I start to propose such an attribute `[MetadataOnly]` to be used on Expression parameters, such that the compiler knows that there is no intention to compile and execute such Expression like the parameter of the OrderBy:
``` CSharp
jobs.OrderBy(j => j.Responsible.Name)
```
In this case the compiler could remove those verification that protect that code from crash on runtime. It will not, because this kind of expression only matters as metadata information in the real world.
Consider this as another kind of forgiveness: "Compiler, forget about runtime here, please"
It is just to start, but again, the language team have a lot more elements, info and knowledge to review all of this better and to find better solutions than single proposals.

**Comment by CyrusNajmabadi** [comment]
> no warning when it cannot be solved please

But it can be solved.  In several different fashions.  Including, but probably not limited to:

1. checking for null
2. using `!`
3. disabling that warnign in that location.
4. disabling NRT for these files or projects.

If you choose to have NRT warnings, then you get warnings when your code appears to be NRT unsafe.  You have many things you can do at that point.  My favorite, personally, is '1' as it appears to be the most sensible option given that the value actually can be null.

**Comment by CyrusNajmabadi** [comment]
> I start to propose such an attribute [MetadataOnly] to be used on Expression parameters, such that the compiler knows that there is no intention to compile and execute such Expression like the parameter of the OrderBy:

Idon't see how this would be helpful.  Would we also not report anything for:

```c#
jobs.OrderBy(j => j.Responsible.NonExistentProperty)
```

?

After all, if there is no intention to compile/execute, then there's no point in doing any checking.  And, if that's the case, why not just use a string?  The purpose of linq is to be able to express your queries *in* the domain of C# using the language constructs *and* semantics of C#.  If hte domain doesn't effectively map to the C# domain, then i think the riht example is to just use a string, or some other mechanism.

**Comment by mariodvm** [comment]
> > no warning when it cannot be solved please
> 
> But it can be solved. In several different fashions. Including, but probably not limited to:
> 
> 1. checking for null
> 2. using `!`
> 3. disabling that warnign in that location.
> 4. disabling NRT for these files or projects.
> 
> If you choose to have NRT warnings, then you get warnings when your code appears to be NRT unsafe. You have many things you can do at that point. My favorite, personally, is '1' as it appears to be the most sensible option given that the value actually can be null.

Of course they could be the current workarounds available for anyone trying to adapt its data layer projects to NRT, but we are not here to suggest workarounds. Isn't it? They are just that: workarounds to not solve anything actually. Let's see:
1. _Checking null_. Ok, we can refactor all code with `exp == null  ? null : ...` that was not needed just before NRT, and that would not be needed anyway if we could use null propagation. This is what null propagation is made for.
2. _Using "!"_ This makes fake code which is not consistent with reality.  I prefer not to patch like that
3. _Disable warning_. This is another unnecessary patch so far (in all over the code).
4. _Disable NRT_. Ok, this is what I'm doing now. Maybe this is what many others are doing too: **choosing not to use nullability on data layer projects at all**. Is it the expected solution? I'm ok with it as workaround (ok, NRT is not 100% aligned with current data layers solutions yet), but I can't settle for this as a definitive solution.

**Comment by mariodvm** [comment]
> > I start to propose such an attribute [MetadataOnly] to be used on Expression parameters, such that the compiler knows that there is no intention to compile and execute such Expression like the parameter of the OrderBy:
> 
> Idon't see how this would be helpful. Would we also not report anything for:
> 
> ```cs
> jobs.OrderBy(j => j.Responsible.NonExistentProperty)
> ```
> 
> ?
> 
> After all, if there is no intention to compile/execute, then there's no point in doing any checking. And, if that's the case, why not just use a string? The purpose of linq is to be able to express your queries _in_ the domain of C# using the language constructs _and_ semantics of C#. If hte domain doesn't effectively map to the C# domain, then i think the riht example is to just use a string, or some other mechanism.

Ok, do you really feel ok with "!" on this kind of code?
```CSharp
jobs.OrderBy(j => j.Responsible!.Name)
```
There are a lot of existing code yet saying just: "hey, order by the [name] of the [responsible]" and not "hey, order by the [name] of the [responsible] assuming that [Responsible] is not null" 
These expressiveness of Expressions have being a fancy way of indicating just metadata (the navigation through property definitions) so far, and it looks broken now with NRT warnings everywhere. 
Or..., maybe we should feel more happy with this...?
```CSharp
jobs.OrderBy($"{nameof(Job.Responsible)}.{nameof(Contact.Name)}")
```
> In that case, it does not apply to `OrderBy` itself as there is not an overload with `string` parameter. Only `Expression<Func<TSource, TKey>>`

**Comment by CyrusNajmabadi** [comment]
> that was not needed just before NRT

Right... 
But htat's the point of NRT. You turn it on, and if your code is not nullsafe... you make it nullsafe.  NRT is effective because it tells you about the potential problem areas, and will understand and stop warning once you've made your code resilient to the issue :)

> Is it the expected solution?

It entirely depends on your use case.  If you're using a library that has different null semantics than C#, then yes, i would expect you would disable this stuff when using that library.  

The point of linq is to model libraries into the c# domain with semantics very close to C#.  If the library deviates, then it's not worthwhile to use it.  For example, say you were evaluating linq in a context where adding integers produced a string, instead of an integer.  You'd have a tough time using that in C#'s linq because C# will always think that `int + int = int` not `int + int = string`.

**Comment by CyrusNajmabadi** [comment]
>Ok, do you really feel ok with "!" on this kind of code?

Yes.  if your end library is not going to have any issues with `null`, then you use `!`.  `!` is exactly the "i know more than what static analysis can ascertain" operator.  That's why it exists in the first place :)

> In that case, it does not apply to OrderBy itself as there is not an overload with string parameter. Only Expression<Func<TSource, TKey>>

You're choosing to use Linq Expression Trees.  Those are intended to map to C# semantics.  I would expect if you didn't want C# semantics to go through some *other* layer :)

> ave being a fancy way of indicating just metadata (the navigation through property definitions) so far, and it looks broken now with NRT warnings everywhere.

Only because *in the metadata* you have stated the values can be `null`.  The reason to state that it can be null is precisely so you can ensure that `the navigation through property definitions` is called out for potentially being unsafe.  If you don't want the warnings about this being unsafe here, then just say that the value is always non-null, and now the compiler will assume that's correct at all these use points, and will not warn you. :)

**Comment by mariodvm** [comment]
> > that was not needed just before NRT
> 
> Right...
> But htat's the point of NRT. You turn it on, and if your code is not nullsafe... you make it nullsafe. NRT is effective because it tells you about the potential problem areas, and will understand and stop warning once you've made your code resilient to the issue :)
> 
> > Is it the expected solution?
> 
> It entirely depends on your use case. If you're using a library that has different null semantics than C#, then yes, i would expect you would disable this stuff when using that library.
> 
> The point of linq is to model libraries into the c# domain with semantics very close to C#. If the library deviates, then it's not worthwhile to use it. For example, say you were evaluating linq in a context where adding integers produced a string, instead of an integer. You'd have a tough time using that in C#'s linq because C# will always think that `int + int = int` not `int + int = string`.

Again...
Maybe I am too wrong but if `Responsible` is a null reference, `Person? Responsible { get; set;}`
 I think the following code means: _Give me all jobs with responsible naming John and don't check if they are null, because I ensure with "!" that all jobs have responsibles!_ Isn't it?
```CSharp
jobs.Where(j => j.Responsible!.Name == "John");
```
Here, it is expected that an exception is thrown if some job has no responsible. But in the real world (so much code out there) it will not throw any exception when using ORMs if some job has no responsible (**INCONSISTENT**). On the contrary it will behavior exactly as if we were using null-propagation like in the following code:
```CSharp
jobs.Where(j => j.Responsible?.Name == "John");
```
which means:  _Give me all jobs with responsible naming John but consider that some jobs have no responsibles_. This is what is actually happening with ORMs and this is exactly what everyone expect in must cases, but this (the most common scenario and what those code is already meaning before NRT) is exactly what is not allowed by prohibiting "?" on Lambda Expressions and giving us warnings instead that can only be really solved with "?" (**INCONSISTENT**)

**Comment by IS4Code** [comment]
Then perhaps effort should be instead directed at pushing expression trees to be finally complete. It's not the fault of nullable reference types that many sensible operations are not working for expression trees, despite having plausible runtime representation. Null propagation, tuples, dynamic invocations, out variables, and the list goes on.

**Comment by mariodvm** [comment]
> Then perhaps effort should be instead directed at pushing expression trees to be finally complete. It's not the fault of nullable reference types that many sensible operations are not working for expression trees, despite having plausible runtime representation. Null propagation, tuples, dynamic invocations, out variables, and the list goes on.

True, and you can choose not to use them because there are not supported... so it means that this is the conclusion: null reference is not really supported for linq, and though data layer models should not be NRT enabled for these projects...

**Comment by iam3yal** [comment]
@mariodvm You can also use the null object pattern so instead of doing the following:
```
public class Job 
{ 
    public Contact? Responsible { get; set; }
}
```
You could do this:
```
public class Job 
{ 
    public Contact Responsible { get; set; }
}
```
and return an null like instance of `Contact` whenever it's null.

**Comment by mariodvm** [comment]
> @mariodvm You can also use the null object pattern so instead of doing the following:
> 
> ```
> public class Job 
> { 
>     public Contact? Responsible { get; set; }
> }
> ```
> 
> You could do this:
> 
> ```
> public class Job 
> { 
>     public Contact Responsible { get; set; }
> }
> ```
> 
> and return an null like instance of `Contact` whenever it's null.

Exactly, this is what we were already doing before NRT. All of this happens if you want to enable NRT to Data Layer and expect no warnings by doing what it is supposed to do with this good new feature.
References were nullables _by default_ before, so old code is warning us if enable NRT. As reference types are _not-nullable_ by default now:
```CSharp
public class Job 
{ 
    public Contact Responsible { get; set; }
}
```
you'll get warning on this:
```CSharp
job.Responsible = null;
```
It's ok, no problem. I can change everything as expected (making Responsible nullable with "?"), also using "?" in query warnings, but... compiler don't allow lambda propagation on expressions so there is a feeling that something is not 100% finished in NRT+Linq. 
The same happens on new warnings raising on every expression intentionally as metadata-only like in the `OrderBy`.

**Comment by CyrusNajmabadi** [comment]
> Here, it is expected that an exception is thrown if some job has no responsible. But in the real world (so much code out there) it will not throw any exception

Then don't mark it with `?`. It doesn't have null semantics then.

**Comment by mariodvm** [comment]
> > Here, it is expected that an exception is thrown if some job has no responsible. But in the real world (so much code out there) it will not throw any exception
> 
> Then don't mark it with `?`. It doesn't have null semantics then.

I said: making `Responsible` nullable with "?", but... Why insisting to solve a single example? It's just an example! The global problem is that some properties have null semantics and some of them have not. **Those with null semantics should be used with "?" with NRT enabled and they should be accessed with null propagation, even in lambda Expression, and they aren't**. The problem is not my example, nor my code. The problem is that the current NRT cannot be applied friendly in so much code (data layers in so many projects) because:
1. That semantics with null propagation on non lambda-expressions are OK, but on lambda expressions are ERRORS. 
2. The nullable condition on properties cannot be obtained by reflection or through some Expression type node... as model builders should not longer need redundant methods like: `IsRequired()`
3. Metadata-only semantic of Expressions, which are used a lot (e.g. OrderBy, Model building routines and Migrations routines) are not distinguished by the compiler kind of errors/warnings.

**Comment by CyrusNajmabadi** [comment]
> That semantics with null propagation on non lambda-expressions are OK, but on lambda expressions are ERRORS.

Yes.  The option there is to use the existing supported expression-tree constructs, or get expression trees to support `?.`.  

> Metadata-only semantic of Expressions, which are used a lot (e.g. OrderBy, Model building routines and Migrations routines) are not distinguished by the compiler kind of errors/warnings.

Correct.  That's how it's been in ExpressionTrees since they were originally introduced.  It's not a different language.  There's no concept of "only apply this subset of rules to them".  i.e. say you're using metadata-only semantics in your domain.  that doesn't mean you get to side-step spelling mistakes in your ExpressionTrees.  Similarly you can't avoid accessibility checks.  Similarly, you don't get a different set of conversion operators between types, etc. etc..  The goal of ExpressionTrees is to allow you to map your *other* domain onto C#.  But you now have to abide by the rules of C# in those trees.

If this is not something you would want, i would recommend disabling NRT analysis around these calls.

**Comment by mariodvm** [comment]
> There's no concept of "only apply this subset of rules to them". 

But maybe it could be 'some different semantics' in the compiler because there is intentionally a different semantics when using them this way. 
There is already different semantics for the same rule in the language. For example: those optional parameters attributed with `[CallerMemberName]` are assigned directly by the compiler when calling  methods with these parameters. This idea can be used to **annotate** Expression Trees parameters in order to be compiled taking this "in mind", so the coder have not to introduce irrelevant `!` on expressions that would never provokes null access in runtime.

**Comment by HaloFour** [comment]
@mariodvm 

Annotation them with what?  And how would that help the compiler know that a particular queryable provider doesn't care about `null` in those cases?  The compiler doesn't know one queryable provider from another.

**Comment by mariodvm** [comment]
> @mariodvm
> 
> Annotation them with what? And how would that help the compiler know that a particular queryable provider doesn't care about `null` in those cases? The compiler doesn't know one queryable provider from another.

With an attribute. The compiler don't need to know anything about what query providers do. `OrderBy` for example, just include a call to itself in the Expression Tree. Then the query provide will do whatever it does; but in any case, this is just a signal made by who is defining such a method like the `OrderBy` to the compiler to avoid further unneeded runtime protections, because he/she knows that this code has no runtime execution.

**Comment by HaloFour** [comment]
@mariodvm 

The LINQ extension methods for expression trees don't know any more about the queryable providers than the compiler does.

**Comment by mariodvm** [comment]
> @mariodvm
> 
> The LINQ extension methods for expression trees don't know any more about the queryable providers than the compiler does.

Exactly, and they don't need it. It is the programmer of the method who knows that he/she is not doing nothing at all with that expression in runtime; nothing except just inspecting the metadata or visiting/modifying the tree, and so on..., so the programmer knows that the expression will not be compiled, nor executed in runtime so those runtime protection are not needed at all. Then, these  parameters could be annotated as [NoRuntime] or [MetadataOnly] ([or whatever]) so the compiler don't have to protect that code against errors that the programmer knows that will no happen at all. In this case is the programmer of the method who is the responsible of that not-compiling-checks, in the same way using `!` is a responsibility of the programmer that tells compiler to forgive some null checks, and also like in `dynamic` where the programmer assumes the responsibility of calling some method that maybe does not exists at all, and the compiler doesn't protect for that because it is what `dynamic` means...
I'm just proposing to include some kind of language construction [an attribute], so the programmer of those methods can tell the compiler to avoid runtime protections on those parameters expressions.

**Comment by HaloFour** [comment]
@mariodvm 

> so the programmer knows that the expression will not be compiled, nor executed in runtime so those runtime protection are not needed at all.

No, the author of those extension methods doesn't know that at all.  All they know is that they are composing on top of an expression tree.  That's no different to composing on top of an `IEnumerable<T>`.  The method itself never executes it, and the method has no idea if/how/when it might be executed.

**Comment by CyrusNajmabadi** [comment]
@mariodvm Can you show an example of what you eman?  i.e. where you would place this attribute, and how the compiler would see it?  That might be helpful for both sides to get the point across.  Thanks!

**Comment by mariodvm** [comment]
> @mariodvm Can you show an example of what you eman? i.e. where you would place this attribute, and how the compiler would see it? That might be helpful for both sides to get the point across. Thanks!

Consider some (sample) method like this
```CSharp
public static IQueryable<T> JoinByNavigation<T,TResult>(
   this IQueryable<T> q, 
   [NoRuntime] Expression<Func<T,Result>> exp)
   => q.Provider.CreateQuery<TSource>(
        Expression.Call(null, 
                        GetMethodInfoOf(() => MyQueryExt.JoinByNavigation(q, exp)),
                        new Expression[] { source.Expression, Expression.Quote(exp) }));

```
so that the underlying query provider would then inspect the tree to find out what tables will be involved in the joins. Then it could be used like this:
```CSharp
myContext.Jobs
         .JoinByNavigation(j => j.Responsible.Address.State)
         .Select(j => new { j.Name, j.Responsible?.Address?.State })
```
> I know this is equivalent to the `Include` and a chain of `ThenInclude` of EF... **It is just an example**

As a result, that null-propagation is not necessary at all when calling to `JoinByNavigation` because it's programmer indicates it is unneeded. In the other hand, I expect that the "same" expression do need lambda-expression propagation in the `Select` (for consistency) because the parameter of Select is not annotated with `[NoRuntime]`
> Note that the proper join avoids null errors later in SQL, but as you said before, the compiler should not rely on those ORM implementation details. It should let us place those lambda propagation in expressions to make it consistent, no matter what ORMs do later.
